### PR TITLE
docs/release.md: Add release documentation

### PR DESCRIPTION
**Pull request by mxinden**
In order to stay consistent across releases and also to enable possibly anyone with the correct access rights to release libp2p, this pull request adds a release documentation to the repository.

I have never cut a release of rust-libp2p. This documentation is based on past discussions and the git commit history. To reviewers: Please ensure this is in line with how you handled releasing for the last couple of versions.

### CODE REVIEW

File: docs/release.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# Release Process
+
+This project follows [semantic versioning](https://semver.org/). The following
+documentation will refer to `X.Y.Z` as _major_, _minor_ and _patch_ version.
+
+## Before the Release
```
Comment (romanb) [inline docs/release.md:6]:
```suggestion
## Development between releases
```

File: docs/release.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  In case there is no `[unreleased]` section yet, create one with an increased
+  major, minor or patch version depending on your change. In addition update the
+  corresponding entry of the crate in the root level `Cargo.toml`.
+
+
+## Cutting a Release
+
+1. For each crate's `CHANGELOG.md` including the top level `libp2p` crate
+   replace `# X.Y.Z [unreleased]` with `# X.Y.Z [yyyy-mm-dd]` and create a
+   combined pull request against the rust-libp2p `master` branch.
```
Comment (romanb) [inline docs/release.md:21]:
```suggestion
1. Set the release date for each crate to be released in the respective `CHANGELOG.md`
   by replacing `# X.Y.Z [unreleased]` with `# X.Y.Z [yyyy-mm-dd]` and create a pull request
   against the rust-libp2p `master` branch.
```

File: docs/release.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+   
+2. Once merged, create and push a tag for each updated crate.
+
+    ```
+    $ tag="<CRATE-NAME>-X.Y.Z"
+    $ git tag -s "${tag}" -m "${tag}"
+    $ git push origin "${tag}"
+    ```
+    
+3. Create and push a tag for the top level `libp2p` crate.
```
Comment (romanb) [inline docs/release.md:31]:
```suggestion
3. Create and push a tag for the top level `libp2p` crate, if it is being released.
```

File: docs/release.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+    ```
+    # Note the additional `v` here.
+    $ tag="vX.Y.Z"
+    $ git tag -s "${tag}" -m "${tag}"
+    $ git push origin "${tag}"
+    ```
+    
+4. Publish each crate including the top level `libp2p` crate to crates.io.
```
Comment (romanb) [inline docs/release.md:40]:
```suggestion
4. Publish each tagged crate to crates.io. `cargo` assists in getting the order of the releases correct.
```

File: docs/release.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+- Every substantial pull request should add an entry to the `[unreleased]`
+  section of the corresponding crate `CHANGELOG.md` file. See
+  [#1698](https://github.com/libp2p/rust-libp2p/pull/1698/files) as an example.
+  
+  In case there is no `[unreleased]` section yet, create one with an increased
+  major, minor or patch version depending on your change. In addition update the
+  corresponding entry of the crate in the root level `Cargo.toml`.
+
+
+## Cutting a Release
```
Comment (romanb) [inline docs/release.md:17]:
```suggestion
## Releasing one or more crates
```

File: docs/release.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+1. Set the release date for each crate to be released in the respective
+   `CHANGELOG.md` by replacing `# X.Y.Z [unreleased]` with
+   `# X.Y.Z [yyyy-mm-dd]` and create a pull request against the rust-libp2p
+   `master` branch.
+
+2. Once merged, create and push a tag for each updated crate.
+
+    ```
+    $ tag="<CRATE-NAME>-X.Y.Z"
+    $ git tag -s "${tag}" -m "${tag}"
```
Comment (twittner) [inline docs/release.md:28]: I have never created a GPG signed tag so far, but went for annotated tags. I also create them all locally and eventually push them altogether  with `git push origin master --tags`.

File: docs/release.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  In case there is no `[unreleased]` section yet, create one with an increased
+  major, minor or patch version depending on your change. In addition update the
+  corresponding entry of the crate in the root level `Cargo.toml`.
+
+
+## Releasing one or more crates
+
+1. Set the release date for each crate to be released in the respective
+   `CHANGELOG.md` by replacing `# X.Y.Z [unreleased]` with
+   `# X.Y.Z [yyyy-mm-dd]` and create a pull request against the rust-libp2p
```
Comment (twittner) [inline docs/release.md:21]: I wonder why we insist on putting those dates in there? The version corresponds to a tag which maps to a commit which has a date. How often do people look at the date in a changelog?

Comment (mxinden) [reply docs/release.md:21]: I am fine either way. I just want to be consistent across all `github.com/libp2p/rust-libp2p` crates. @romanb thoughts?

File: docs/release.md (hunk 6 above)
Comment (mxinden) [reply docs/release.md:29]: Added a comment above. Let me know if that looks good to you @twittner.

File: docs/release.md (hunk 7 above)
Comment (romanb) [reply docs/release.md:22]: I don't care about the dates in the changelogs much. That tradition was started [here](https://github.com/libp2p/rust-libp2p/pull/1614#discussion_r442206372).

**Comment by mxinden** [comment]
@twittner https://github.com/libp2p/rust-libp2p/pull/1712/commits/17eda9542f065ffae6e7d460699cf033a9e842f2 removes the requirement to include release dates in the changelog files.

Any other comments / suggestions?

**Comment by romanb** [comment]
:+1:  Let's start with this. Thanks for the writeup.

**Pull request merged**